### ISSUE TITLE
extract_features

### ISSUE THREAD

**Issue by chen849157649** [guoday/ICME2019-CTR#4]
郭大，我最近在学习你的工程代码，目前遇到几点疑问？烦请解答
在计算title_mean时，出现   
mean=0        # {"item_id": 1488899, "title_features": {"314": 1, "3499": 1, "5285": 1, "174": 1}}
 for t1,t2 in zip(item[0].split(),item[1]):
            mean+=int(t1)*t2
  mean/=sum(item[1])
tract2_title.txt文件中的title_features的key表示的是什么？value应该是词频吧，为啥你能将key和value相乘，然后去求mean，这里不理解。

在var_mean部分，你使用了傅里叶变换，用它的作用或者意义是啥呢？
data=train_df[['uid','author_id']].append(test_df[['uid','author_id']])
 group=data[['uid','author_id']].groupby('uid')
 group=group.apply(lambda x:np.var(np.fft.fft(list(Counter(list(x['author_id'])).values()))))
 train_df['uid_num_of_author_fft_var']=train_df['uid'].apply(lambda x:group[x])
 test_df['uid_num_of_author_fft_var']=test_df['uid'].apply(lambda x:group[x])
 print(train_df[train_df['finish']==1]['uid_num_of_author_fft_var'].mean())
 print(train_df[train_df['finish']==0]['uid_num_of_author_fft_var'].mean())

**Comment by guoday** [comment]
> 郭大，我最近在学习你的工程代码，目前遇到几点疑问？烦请解答
> 在计算title_mean时，出现
> mean=0 # {"item_id": 1488899, "title_features": {"314": 1, "3499": 1, "5285": 1, "174": 1}}
> for t1,t2 in zip(item[0].split(),item[1]):
> mean+=int(t1)*t2
> mean/=sum(item[1])
> tract2_title.txt文件中的title_features的key表示的是什么？value应该是词频吧，为啥你能将key和value相乘，然后去求mean，这里不理解。
> 
> 在var_mean部分，你使用了傅里叶变换，用它的作用或者意义是啥呢？
> data=train_df[['uid','author_id']].append(test_df[['uid','author_id']])
> group=data[['uid','author_id']].groupby('uid')
> group=group.apply(lambda x:np.var(np.fft.fft(list(Counter(list(x['author_id'])).values()))))
> train_df['uid_num_of_author_fft_var']=train_df['uid'].apply(lambda x:group[x])
> test_df['uid_num_of_author_fft_var']=test_df['uid'].apply(lambda x:group[x])
> print(train_df[train_df['finish']==1]['uid_num_of_author_fft_var'].mean())
> print(train_df[train_df['finish']==0]['uid_num_of_author_fft_var'].mean())

对于第一个特征，其实是针对于这个数据集的，因为这个数据集，好像是按某种方式编码的，所以求出词的中位数会有点效果。而对于第二个，算出每个用户点击不同作者的广告次数，如三个广告，分别点击次数[4,2,6]，我们可以求中位数，均值，方差等来表示这个list，但用傅立叶变换转成频域，然后求均值也是能够刻画这个list的。

**Comment by chen849157649** [comment]
@guoday ，谢谢，郭大，我训练了一遍，track2的数据，finish task AUC=0.736733， like task是否是将前面的数据处理部分的finish 替换为like部分，重新prepocess, extract_features, convert_format吗？

**Comment by guoday** [comment]
> @guoday ，谢谢，郭大，我训练了一遍，track2的数据，finish task AUC=0.736733， like task是否是将前面的数据处理部分的finish 替换为like部分，重新prepocess, extract_features, convert_format吗？


> @guoday ，谢谢，郭大，我训练了一遍，track2的数据，finish task AUC=0.736733， like task是否是将前面的数据处理部分的finish 替换为like部分，重新prepocess, extract_features, convert_format吗？

不需要，run.py 中的label='finish'设为label='like'

**Comment by chen849157649** [comment]
@guoday ，好的，我之前做视频理解及图像方面的，对推荐方面是刚开始接触，所以问题比较多，谢谢你的细心解答。
明白了，这个label就是真实标签，数据处理部分确实 特征交叉处理上finish，like都有考虑，按你说的将label='finish'替换训练，那这样的话，其实就是多目标排序模型，要推理的话，模型需要保存两次权重参数了，根据目标选择加载哪个权重，另外工程中没有保存模型权重参数，我加上保存模型部分后，另外再写个inference.py脚本，包装成服务接口后，这主体部分一直迭代数据训练，推理服务定时更新即可预测了吧，我这样理解，是否正确合理？

**Comment by guoday** [comment]
> 实就是多目标排序模型

我不确定一直迭代训练性能会不会变差，毕竟我还没试过线上推荐。至于保存模型权重，然后读取推理的话，可能在我这个code有点难度，因为我用的是hash编码，这个hash的随机种子会在你下一次运行程序时发生变化，比如hash('a')=2,下一次运行程序时有可能hash('a')!=2。不过你可以在src/mistic_utlis.py 中定义新的hash函数
def hash(key,value,vocab_num):
    s='key_'+str(key)+' value_'+str(value)
    seed=131
    h=0
    for i in s:
        h = (h * seed) + ord(i)
    return abs(h)%vocab_num

**Comment by chen849157649** [comment]
@guoday ，恩，明白了。工程代码数据处理部分全部读完了，训练部分还没看完。边看边感叹你真是好厉害，特征之间交叉构造以及网络搭建等等，代码质量都好棒。刚才看了你说那部分。因为是采用hash对single_features编码的，每次重新训练，数据的编码表示都有变化，所以按照你说的重新定义替换后，同样的字段数据编码不会改变，模型保存参数后具有推理能力，修改如下，
![微信截图_20200605141840](https://user-images.githubusercontent.com/32058220/83843839-6d2e9d00-a738-11ea-870f-42f74fbc2b5e.png)

**Comment by guoday** [comment]
> @guoday ，恩，明白了。工程代码数据处理部分全部读完了，训练部分还没看完。边看边感叹你真是好厉害，特征之间交叉构造以及网络搭建等等，代码质量都好棒。刚才看了你说那部分。因为是采用hash对single_features编码的，每次重新训练，数据的编码表示都有变化，所以按照你说的重新定义替换后，同样的字段数据编码不会改变，模型保存参数后具有推理能力，修改如下，
> ![微信截图_20200605141840](https://user-images.githubusercontent.com/32058220/83843839-6d2e9d00-a738-11ea-870f-42f74fbc2b5e.png)

https://github.com/guoday/ICME2019-CTR/blob/dbc51b0f800c3772da6e15d44ff6e989bc357845/src/misc_utils.py#L38
还有这个

**Comment by chen849157649** [comment]
@guoday ，知道，你使用hash这种编码，产生的其实就是weights embedding中的索引，
dnn_input.append(tf.reshape(tf.gather(self.single_emb_v2, self.single_features),[-1,len(hparams.single_features)*hparams.single_k]))，那如果是one-hot编码形式，然后记录特征索引采用这三行代码
self.embeddings = tf.nn.embedding_lookup(self.weights['feature_embeddings'],self.feat_index)  # N * F * K，
feat_value = tf.reshape(self.feat_value,shape=[-1,self.field_size,1])
self.embeddings = tf.multiply(self.embeddings,feat_value)
效果与上面的那种方式一样了吧。
